File: desktop/src/lib/bridge/protocol.ts
export type BridgeEventName =
  | "agent.turn.started"
  | "agent.text.delta"
  | "agent.reasoning.delta"
  | "agent.tool.call"
  | "agent.tool.result"
  | "agent.error"
  | "agent.turn.finish"
  | "permission.request"
  | "session.updated";

export type BridgeEvent = {
  type: "event";
  event: BridgeEventName;
  sessionId: string;
  turnId?: number;
  payload: Record<string, unknown>;
};

export type BridgeRequestMethod =
  | "app.bootstrap"
  | "session.archive"
  | "session.create"
  | "session.resume"
  | "git.checkout"
  | "model.select"
  | "model.configure"
  | "turn.start"
  | "turn.interrupt"
  | "permission.resolve";

export type BridgeRequest = {
  type: "request";
  id: string;
  method: BridgeRequestMethod;
  token: string;
  payload: Record<string, unknown>;
};

export type BridgeResponse = {
  type: "response";
  id: string;
  ok: boolean;
  payload?: Record<string, unknown>;
  error?: {
    code: string;
    message: string;
  };
};

export type TurnStartResult = {
  sessionId: string;
  turnId: number;
  finishReason: string;
};

export type BootstrapSessionSummary = {
  id: string;
  title: string;
  isActive?: boolean;
  isEmpty?: boolean;
};

export type BootstrapSessionGroup = {
  id: string;
  folderName: string;
  workspacePath?: string;
  sessions: BootstrapSessionSummary[];
};

export type BootstrapResult = {
  sessionId: string;
  sessionGroups: BootstrapSessionGroup[];
  workspacePath: string;
  branch: string;
  branches: string[];
  isDirty: boolean;
  activeProfile: string;
  modelProfiles: ModelProfileSummary[];
};

export type ModelProfileSummary = {
  id: string;
  label: string;
  modelName: string;
  source?: "builtin" | "custom";
  baseUrl?: string;
  hasApiKey?: boolean;
};

export type ModelConfigInput = {
  id?: string;
  displayName: string;
  baseUrl: string;
  apiKey: string;
  modelName: string;
};

export type VoiceSettings = {
  provider: "custom" | "stepfun";
  sttUrl: string;
  ttsUrl: string;
  stepfunVoice: string;
  stepfunKey: string;
  hasStepfunKey: boolean;
  ttsEnabled: boolean;
};

export type VoiceSettingsInput = {
  provider: "custom" | "stepfun";
  sttUrl: string;
  ttsUrl: string;
  stepfunVoice: string;
  stepfunKey: string;
  ttsEnabled: boolean;
};

export type ResolvedModelConfig = ModelConfigInput & { id: string };

export type RestoredConversation = {
  turns: Array<{
    id: number;
    sessionId: string;
    userText: string;
    assistantText: string;
    reasoningText: string;
    status: "streaming" | "waiting_permission" | "completed" | "error";
    finishReason: string;
    tools: Array<{
      id: string;
      name: string;
      args: Record<string, unknown>;
      result: string;
      resultPreview: string;
      status: "running" | "completed" | "error";
    }>;
    errorText: string;
  }>;
  activeTurnId: number | null;
};

export type SessionResumeResult = {
  sessionId: string;
  workspacePath: string;
  branch: string;
  branches: string[];
  isDirty: boolean;
  conversation: RestoredConversation;
};

export type TurnStartOptions = {
  permissionMode: string;
};

export type PermissionResolveBehavior = "allow" | "deny";
export type PermissionRememberScope = "" | "session" | "workspace";

export type PermissionResolvePayload = {
  requestId: string;
  behavior: PermissionResolveBehavior;
  rememberScope: PermissionRememberScope;
  message?: string;
};

export type DesktopBridgeClient = {
  connect: () => Promise<void>;
  bootstrap: () => Promise<BootstrapResult>;
  archiveSession: (sessionId: string) => Promise<boolean>;
  createSession: () => Promise<{ sessionId: string; workspacePath: string }>;
  resumeSession: (sessionId: string) => Promise<SessionResumeResult>;
  checkoutBranch: (branch: string) => Promise<{ branch: string; branches: string[]; isDirty: boolean }>;
  selectModelProfile: (profile: string) => Promise<{ profile: string; modelName: string }>;
  configureModel: (config: ResolvedModelConfig) => Promise<{ profile: string; modelName: string }>;
  saveModelConfig: (input: ModelConfigInput) => Promise<ModelProfileSummary>;
  deleteModelConfig: (id: string) => Promise<boolean>;
  loadVoiceSettings: () => Promise<VoiceSettings>;
  saveVoiceSettings: (input: VoiceSettingsInput) => Promise<VoiceSettings>;
  startTurn: (text: string, options?: TurnStartOptions) => Promise<TurnStartResult>;
  interruptTurn: (turnId?: number) => Promise<boolean>;
  resolvePermission: (payload: PermissionResolvePayload) => Promise<boolean>;
  close: () => void;
};

export function readString(payload: Record<string, unknown>, key: string): string {
  const value = payload[key];
  return typeof value === "string" ? value : "";
}

export function readRecord(payload: Record<string, unknown>, key: string): Record<string, unknown> {
  const value = payload[key];
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return {};
}

export function readNumber(payload: Record<string, unknown>, key: string): number {
  const value = payload[key];
  return typeof value === "number" ? value : 0;
}

export function readBoolean(payload: Record<string, unknown>, key: string): boolean {
  const value = payload[key];
  return typeof value === "boolean" ? value : false;
}

export function isBridgeEvent(value: unknown): value is BridgeEvent {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value as Partial<BridgeEvent>;
  return candidate.type === "event" && typeof candidate.event === "string";
}

export function isBridgeResponse(value: unknown): value is BridgeResponse {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value as Partial<BridgeResponse>;
  return candidate.type === "response" && typeof candidate.id === "string";
}

File: src/voice_code/desktop/protocol.py
"""Typed protocol objects for the local desktop bridge."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from voice_code.agent.types import AgentEvent, EventType
from voice_code.permissions import PermissionRequest
from voice_code.security import redact_secrets

BridgeEventName = Literal[
    "agent.turn.started",
    "agent.text.delta",
    "agent.reasoning.delta",
    "agent.tool.call",
    "agent.tool.result",
    "agent.error",
    "agent.turn.finish",
    "permission.request",
    "session.updated",
]

PermissionResolveBehavior = Literal["allow", "deny"]
PermissionRememberScope = Literal["", "session", "workspace"]
BridgeEmitter = Callable[["BridgeEvent"], None | Awaitable[None]]
BridgeRequestMethod = Literal[
    "app.bootstrap",
    "session.archive",
    "session.create",
    "session.resume",
    "git.checkout",
    "model.select",
    "model.configure",
    "turn.start",
    "turn.interrupt",
    "permission.resolve",
]


class BridgeEvent(BaseModel):
    """A stable event envelope consumed by the desktop frontend."""

    model_config = ConfigDict(populate_by_name=True)

    type: Literal["event"] = "event"
    event: BridgeEventName
    session_id: str = Field(alias="sessionId")
    turn_id: int | None = Field(default=None, alias="turnId")
    payload: dict[str, object] = Field(default_factory=dict)


class TurnStartResult(BaseModel):
    """Result returned once a bridge turn has completed."""

    session_id: str
    turn_id: int
    finish_reason: str = ""


class PermissionResolvePayload(BaseModel):
    """UI decision for a pending permission request."""

    model_config = ConfigDict(populate_by_name=True)

    request_id: str = Field(alias="requestId")
    behavior: PermissionResolveBehavior
    remember_scope: PermissionRememberScope = Field(default="", alias="rememberScope")
    message: str = ""


class BridgeRequest(BaseModel):
    """A request sent by a desktop client over any transport."""

    type: Literal["request"] = "request"
    id: str
    method: BridgeRequestMethod
    token: str
    payload: dict[str, object] = Field(default_factory=dict)


class BridgeError(BaseModel):
    code: str
    message: str


class BridgeResponse(BaseModel):
    """A stable response envelope for desktop bridge requests."""

    type: Literal["response"] = "response"
    id: str = ""
    ok: bool
    payload: dict[str, object] | None = None
    error: BridgeError | None = None


def map_agent_event(session_id: str, event: AgentEvent) -> BridgeEvent:
    """Map internal AgentEvent values to stable desktop event names."""
    if event.type == EventType.TEXT:
        return BridgeEvent(
            event="agent.text.delta",
            session_id=session_id,
            turn_id=event.turn,
            payload={"content": event.content},
        )
    if event.type == EventType.REASONING:
        return BridgeEvent(
            event="agent.reasoning.delta",
            session_id=session_id,
            turn_id=event.turn,
            payload={"content": event.content},
        )
    if event.type == EventType.TOOL_CALL:
        return BridgeEvent(
            event="agent.tool.call",
            session_id=session_id,
            turn_id=event.turn,
            payload={
                "toolCallId": event.tool_call_id,
                "toolName": event.tool_name,
                "toolArgs": _safe_payload(event.tool_args),
            },
        )
    if event.type == EventType.TOOL_RESULT:
        return BridgeEvent(
            event="agent.tool.result",
            session_id=session_id,
            turn_id=event.turn,
            payload={
                "toolCallId": event.tool_call_id,
                "toolName": event.tool_name,
                "toolResult": event.tool_result,
                "resultPreview": _preview_text(event.tool_result),
            },
        )
    if event.type == EventType.ERROR:
        if event.tool_call_id:
            return BridgeEvent(
                event="agent.tool.result",
                session_id=session_id,
                turn_id=event.turn,
                payload={
                    "toolCallId": event.tool_call_id,
                    "toolName": event.tool_name,
                    "toolResult": event.content,
                    "resultPreview": _preview_text(event.content),
                    "status": "error",
                },
            )
        return BridgeEvent(
            event="agent.error",
            session_id=session_id,
            turn_id=event.turn,
            payload={
                "content": event.content,
                "status": event.status,
                "errorCode": event.error_code,
            },
        )
    if event.type == EventType.FINISH:
        return BridgeEvent(
            event="agent.turn.finish",
            session_id=session_id,
            turn_id=event.turn,
            payload={
                "finishReason": event.finish_reason,
                "content": event.content,
            },
        )
    return BridgeEvent(
        event="agent.error",
        session_id=session_id,
        turn_id=event.turn,
        payload={"content": f"Unsupported agent event: {event.type.name}"},
    )


def make_turn_started_event(*, session_id: str, turn_id: int, user_text: str) -> BridgeEvent:
    return BridgeEvent(
        event="agent.turn.started",
        session_id=session_id,
        turn_id=turn_id,
        payload={"userText": user_text},
    )


def make_permission_request_event(
    *,
    session_id: str,
    request_id: str,
    request: PermissionRequest,
) -> BridgeEvent:
    return BridgeEvent(
        event="permission.request",
        session_id=session_id,
        payload={
            "requestId": request_id,
            "toolName": request.tool_name,
            "toolInput": _safe_payload(request.tool_input),
            "reason": request.reason,
            "isDestructive": request.is_destructive,
            "riskCategory": request.risk_category,
            "reasonType": request.reason_type,
            "matchedRule": request.matched_rule,
            "ruleSource": request.rule_source,
            "taskId": request.task_id or "",
            "agentType": request.agent_type or "",
        },
    )


def _safe_payload(payload: dict[str, object]) -> dict[str, object]:
    redacted = redact_secrets(payload)
    return redacted if isinstance(redacted, dict) else {}


def _preview_text(text: str, *, limit: int = 160) -> str:
    normalized = _clean_tool_use_error(" ".join(text.split()))
    if len(normalized) <= limit:
        return normalized
    return normalized[: limit - 1] + "…"


def _clean_tool_use_error(text: str) -> str:
    return text.replace("<tool_use_error>", "").replace("</tool_use_error>", "")
